# Ligase Fidelity Profiling

This is an official repository for computational workflow on profiling ligase fidelity using PacBio SMRT sequencing. Please see citing information below.

# Prerequisites

The computational workflow requires a number of tools to be installed and available from the command line in your system.

## PacBio tools

* ```lima``` (optional) - Demultiplexing PacBio data. Available as part of [pbbioconda](https://github.com/PacificBiosciences/pbbioconda).
* ```pbmm2``` - minimap2 SMRT wrapper for PacBio data. Available as part of [pbbioconda](https://github.com/PacificBiosciences/pbbioconda).
* ```ccs``` - Generating circular consensus sequences (CCS) based on PacBio sequencing data. Available as part of [pbbioconda](https://github.com/PacificBiosciences/pbccs).

## Third-party tools

* ```samtools``` - Handling high-throughput sequencing data. Available through [conda](https://anaconda.org/bioconda/samtools) or [GitHub](https://github.com/samtools/samtools).

## Custom PERL scripts

These custom scripts are provided as part of this GitHub repository. The scripts must be made available from the command line in your system (for example, by adding the scripts directory to ```$PATH``` environment variable).

* ```cluster.pl``` - Identify top/bottom subreads in PacBio sequencing reads.
* ```split.pl``` - Split top/bottom subreads to separate files.
* ```extract.pl``` - ???Extract overhangs and barcodes.
* ```bam2csv.pl``` - ???
* ```reporter.pl``` - ???
* ```mktable_barcode.pl``` - Compute nucleotide frequencies in barcodes.
* ```mktable.pl``` - Tabulate ligase fidelity results.

# Computational workflow

# Citing

* Potapov V, Ong JL, Kucera RB, Langhorst BW, Bilotti K, Pryor JM, Cantor EJ, Canton B, Knight TF, Evans Jr TC, and Lohman GJS (2018). Comprehensive Profiling of Four Base Overhang Ligation Fidelity by T4 DNA Ligase and Application to DNA Assembly. *ACS Synth. Biol.* 7(11):2665â€“2674. https://doi.org/10.1021/acssynbio.8b00333

<!-- 


# Dependencies

## Global environments
```
conda activate pbbioconda
module load samtools-1.3.1
```

```
qsub \
        -v root="$PROJDIR",rundir="$rundir",rname="$reference",collectionPathUri="$collectionPathUri",instrument="SEQUEL" \
        -N "demux$sampleId" \
        -o "$rundir"/workflow.log \
        -j yes \
        "$PROJDIR"/scripts/workflow-demux.sh
```

## Perl scripts
```
### cluster reads
"$root"/bin/cluster.pl

### split forward and reverse reads
"$root"/bin/split.pl

### extract barcodes and overhangs
"$root"/bin/extract.pl

### extract results
"$root"/bin/reporter.pl

### barcode table
"$root"/bin/mktable_barcode.pl

### result tables
"$root"/bin/mktable.pl

### optional
"$root"/bin/split_overhangs_by_base.pl
/mnt/home/potapov/projects/160202.LigaseFidelity/blunt_matrix_2.pl
```

## PacBio tools
```
samtools
ccs
blasr
```



## Timings
```
Cluster reads

real    65m3.296s
user    106m29.048s
sys     5m41.955s
```

```
Split reads

real    54m20.970s
user    78m1.393s
sys     11m27.359s
```

```
SAM-to-BAM

real    15m20.012s
user    13m40.762s
sys     1m2.839s
```
# CP-LigaseFidelity -->
